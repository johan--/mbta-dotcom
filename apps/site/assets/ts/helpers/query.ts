export interface QueryParams {
  [key: string]: string;
}

export const paramsToString = (params: QueryParams): string => {
  const p = new URLSearchParams("");
  Object.entries(params).forEach(([key, value]) => {
    p.set(key, value);
  });
  return params ? `?${p.toString()}` : "";
};

export const parseQuery = (query: string): QueryParams => {
  const p = new URLSearchParams(query);

  const allParams: QueryParams = {};
  p.forEach((value, key) => {
    allParams[key] = value;
  });

  return allParams;
};
