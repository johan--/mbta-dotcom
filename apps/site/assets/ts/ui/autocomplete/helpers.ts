import {
  MBTAItemType,
  ContentItem,
  RouteItem,
  SourceItem,
  StopItem
} from "./__autocomplete";

function isStopItem(x: SourceItem): x is StopItem {
  return Object.keys(x).includes("stop");
}

export function isRouteItem(x: SourceItem): x is RouteItem {
  return Object.keys(x).includes("route");
}

export function isContentItem(x: SourceItem): x is ContentItem {
  return Object.keys(x).includes("_content_type");
}

/**
 * @param {SourceItem} item - A result retreived from the autocomplete
 * @returns The surmised type of data represented in the `item`.
 */
export const getItemType = (item: SourceItem): MBTAItemType => {
  if (isStopItem(item)) return "stops";
  if (isRouteItem(item)) return "routes";
  if (isContentItem(item)) return "drupal";
  return "locations";
};

/**
 * @param {SourceItem} item - A result retreived from the autocomplete
 * @returns The path of the attribute to be highlighted in a
 * `AutocompleteComponents` type component.
 */
export const getTitleAttribute = (item: SourceItem): string[] => {
  if (isStopItem(item)) {
    return ["stop", "name"];
  }
  if (isRouteItem(item)) {
    return ["route", "name"];
  }

  // @ts-ignore
  // eslint-disable-next-line no-underscore-dangle
  return Object.keys(item._highlightResult);
};
