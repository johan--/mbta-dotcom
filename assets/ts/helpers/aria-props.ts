export type AriaLabelling = AriaLabel | AriaLabelledBy;

export interface AriaLabel {
  label: string;
}

export interface AriaLabelledBy {
  elementId: string;
}

interface AriaLabelProp {
  "aria-label": string;
}

interface AriaLabelledByProp {
  "aria-labelledby": string;
}

const isLabel = (label: AriaLabelling): label is AriaLabel =>
  (label as AriaLabel).label !== undefined;

export const labelOrDescribedBy = (
  label: AriaLabel | AriaLabelledBy
): AriaLabelProp | AriaLabelledByProp => {
  if (isLabel(label)) {
    return { "aria-label": label.label };
  }
  return { "aria-labelledby": label.elementId };
};
