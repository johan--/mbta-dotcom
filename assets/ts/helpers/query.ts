export interface QueryParams {
  [key: string]: string;
}

// The encoder/decoder functions must be injected
// because this helper is used by some vanilla JS files,
// and MochaJS insists that the window object does not
// include encodeURIComponent or decodeURIComponent,
// which causes a compiler error. If/when there are
// no more JS modules using these helpers, the
// injections should be removed.
type EncoderDecoder = (str: string) => string;

export const paramsToString = (
  params: QueryParams,
  encoder: EncoderDecoder
): string => {
  if (!params || Object.keys(params).length === 0) {
    return "";
  }

  return `?${Object.keys(params)
    .map(key => `${key}=${encoder(params[key])}`)
    .join("&")}`;
};

type QueryReducer = (params: QueryParams, kvString: string) => QueryParams;

const queryReducer = (decoder: EncoderDecoder): QueryReducer => (
  params: QueryParams,
  kvString: string
) => {
  const [key, val] = kvString.split("=");
  return { ...params, [decoder(key)]: decoder(val.replace(/\+/g, " ")) };
};

export const parseQuery = (
  query: string,
  decoder: EncoderDecoder
): QueryParams =>
  query
    ? query
        .substring(1)
        .split("&")
        .reduce(queryReducer(decoder), {})
    : {};
