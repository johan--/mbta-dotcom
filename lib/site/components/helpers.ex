defmodule Site.Components.Helpers do
  @doc """
    Takes a string, capitalizes each word between underscores, and removes the underscores
  """
  @spec alias_name(atom) :: String.t()
  def alias_name(atom) do
    atom
    |> Kernel.to_string()
    |> String.split("_")
    |> Enum.map(&String.capitalize(&1))
    |> Enum.join("")
  end

  @doc """
    A helper function to ensure that the correct path is returned no
    matter where this module is called from (i.e. using mix test or iex)
  """
  @spec components_folder_path :: String.t()
  def components_folder_path do
    File.cwd!()
    |> String.split("/lib/site")
    |> List.first()
    |> Path.join("/lib/site/components/")
  end

  def components_section_path(section) when is_atom(section) do
    components_folder_path()
    |> Path.join("#{section}")
  end

  def component_folder_path(component, section) when is_atom(component) and is_atom(section) do
    section
    |> components_section_path()
    |> Path.join("/")
    |> Path.join("#{component}")
  end

  @doc """
    Returns the component's module (i.e. Site.Components.Buttons.ModeButtonList)
  """
  @spec component_module(atom, atom) :: module
  def component_module(component, section) when is_atom(component) and is_atom(section) do
    Site.Components
    |> Module.concat(alias_name(section))
    |> Module.concat(alias_name(component))
  end
end
