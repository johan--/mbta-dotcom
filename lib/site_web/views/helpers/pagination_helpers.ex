defmodule SiteWeb.PaginationHelpers do
  @moduledoc """
  Pagination generation component that:
  - handles link generation passing various params back to the server
  - shows next / previous buttons
  - shows 1... or ...{last} when the user has many results and they have navigated towards the middle
  - shows 5 middle links on desktop
  - shows 3 middle links for mobile
  """
  use SiteWeb, :view

  @type link_context_t :: %{form: String.t(), path: String.t(), params: map}

  @spec build_link(link_context_t) :: String.t()
  defp build_link(%{path: path, form: form, params: params}) do
    "#{path}?_utf8=âœ“#{query_params(form, Map.put(params, "[offset]", "OFFSET"))}"
  end

  @spec query_params(String.t(), map) :: String.t()
  defp query_params(form, params) do
    Enum.reduce(params, "", fn {key, value}, acc ->
      "#{acc}&#{URI.encode_query(%{(form <> key) => value})}"
    end)
  end

  @spec offset_link(String.t(), integer) :: String.t()
  defp offset_link(link, offset), do: String.replace(link, "OFFSET", to_string(offset - 1))
end
