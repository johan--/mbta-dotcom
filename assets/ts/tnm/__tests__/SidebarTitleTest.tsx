import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import SidebarTitle from "../components/SidebarTitle";
import { createReactRoot } from "../../app/helpers/testUtils";

it("it renders", () => {
  createReactRoot();

  const tree = renderer
    .create(<SidebarTitle dispatch={() => {}} viewType="Routes" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("it changes the view when change view is clicked", () => {
  const mockDispatch = jest.fn();

  const wrapper = shallow(
    <SidebarTitle viewType="Routes" dispatch={mockDispatch} />
  );

  wrapper.find(".m-tnm-sidebar__view-change").simulate("click");
  expect(mockDispatch).toHaveBeenCalledWith({
    type: "CLICK_VIEW_CHANGE",
    payload: { stopId: null }
  });
});

it("it changes the view when change view is selected via ENTER", () => {
  const mockDispatch = jest.fn();

  const wrapper = shallow(
    <SidebarTitle viewType="Routes" dispatch={mockDispatch} />
  );

  wrapper
    .find(".m-tnm-sidebar__view-change")
    .simulate("keyPress", { key: "Enter" });
  expect(mockDispatch).toHaveBeenCalledWith({
    type: "CLICK_VIEW_CHANGE",
    payload: { stopId: null }
  });
});
