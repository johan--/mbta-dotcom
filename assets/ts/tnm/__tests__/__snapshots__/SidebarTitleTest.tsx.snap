// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<div
  className="m-tnm-sidebar__header"
>
  <div
    className="m-tnm-sidebar__header-title"
  >
    <h2>
      Nearby Routes
    </h2>
    <div
      className="m-tnm-sidebar__view-change"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      Change view
    </div>
  </div>
</div>
`;
